src/app_engine/probers.py

```diff
@@ -94,18 +96,64 @@ def get(self):
     self.handle_ceod_response(error_message, status_code)
 
 class ProbeColliderPage(webapp2.RequestHandler):
-  def handle_collider_response(self, error_message, status_code):
-    self.response.set_status(status_code)
+  def handle_collider_response(
+      self, error_message, status_code, collider_instance):
+    result = {
+        constants.WSS_HOST_STATUS_CODE_KEY: status_code
+    }
     if error_message is not None:
-      send_alert_email('Collider Error', error_message)
+      send_alert_email('Collider Error (' + collider_instance + ')',
+                       error_message)
 
       logging.warning('Collider prober error: ' + error_message)
-      self.response.out.write(error_message)
+      result[constants.WSS_HOST_ERROR_MESSAGE_KEY] = error_message
+      result[constants.WSS_HOST_IS_UP_KEY] = False
     else:
-      self.response.out.write('Success!')
+      result[constants.WSS_HOST_IS_UP_KEY] = True
+    return result
+
+  def store_instance_state(self, probing_results):
+    # Store a list with status for each collider host.
+    # The list should be in order of decreasing uptime (the first server in
+    # the list has been up the longest).
+    memcache_client = memcache.Client()
+    for retries in xrange(constants.MEMCACHE_RETRY_LIMIT):
+      active_host = memcache_client.gets(constants.WSS_HOST_ACTIVE_HOST_KEY)
+      if active_host is None:
+        memcache_client.set(constants.WSS_HOST_ACTIVE_HOST_KEY, '')
+        active_host = memcache_client.gets(constants.WSS_HOST_ACTIVE_HOST_KEY)
+      active_host = self.create_collider_active_host(active_host,
+                                                     probing_results)
+      if memcache_client.cas(constants.WSS_HOST_ACTIVE_HOST_KEY, active_host):
+        logging.info('collider status saved to memcache')
+        break
+      logging.warning('retry # ' + str(retries) + ' to set collider status')
+
+  def create_collider_active_host(self, old_active_host, probing_results):
+    # If the old_active_host is still up, keep it. If not, pick a new active
+    # host that is up.
+    try:
+      if (old_active_host in probing_results and
```

Comment:
If it's bad data it means we wrote code wrong somehow and should trigger an assert / throw an exception. That is to say, we should try/catch expected exceptions, but we should let unexpected exceptions trigger. This should be storing the result from handle_collider_response, which we know is a dictionary for sure (or at least it should be). Using get protects us if for some reason the key doesn't exist (through it should since we always set it above).